File: services/llm-gateway/src/llm_gateway/api/health.py
import logging

from fastapi import APIRouter, HTTPException, Request

from llm_gateway.db.postgres import acquire_connection, missing_table_privileges

logger = logging.getLogger(__name__)

health_router = APIRouter()


@health_router.get("/")
async def root() -> dict[str, str]:
    return {"service": "llm-gateway", "status": "running"}


@health_router.get("/_readiness")
async def readiness(request: Request) -> dict[str, str]:
    try:
        async with acquire_connection(request.app.state.db_pool) as conn:
            missing = await missing_table_privileges(conn)
    except Exception:
        logger.exception("Readiness check failed: database connection error")
        raise HTTPException(status_code=503, detail="Database not ready") from None
    if missing:
        # Runs on every probe: a revoked grant unreadies serving pods too,
        # deliberately, since every declared table is on the auth path. The
        # body matches the connection-failure branch (the probe may be
        # public); this log line carries the table names.
        logger.error(
            "Readiness check failed: role lacks SELECT on %s; grants are the users.tf allowlist in posthog-cloud-infra",
            ", ".join(missing),
        )
        raise HTTPException(status_code=503, detail="Database not ready")
    return {"status": "ready"}


@health_router.get("/_liveness")
async def liveness() -> dict[str, str]:
    return {"status": "alive"}

File: services/llm-gateway/internal/server/health.go
package server

import (
	"context"
	"net/http"
	"time"
)

func (a *App) root(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, 200, map[string]string{"service": "llm-gateway", "status": "running"})
}

func (a *App) liveness(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, 200, map[string]string{"status": "alive"})
}

func (a *App) readiness(w http.ResponseWriter, r *http.Request) {
	ctx, cancel := context.WithTimeout(r.Context(), 2*time.Second)
	defer cancel()
	if err := a.db.Ping(ctx); err != nil {
		writeJSON(w, 503, map[string]string{"detail": "Database not ready"})
		return
	}
	writeJSON(w, 200, map[string]string{"status": "ready"})
}
